main/src/library/MutList.flix

```diff
@@ -18,145 +18,407 @@
+        if (i >= n)
+            ()
+        else {
+            a[i] = f(a[i], i);
+            transformWithIndexHelper!(f, a, i + 1, n)
+        }
+
+    ///
+    /// Alias for `foldLeft`.
+    ///
+    pub def fold(f: (b, a) -> b & e, s: b, v: MutList[a]): b & Impure = foldLeft(f, s, v)
 
-    // TODO: reduceLeft
-    // TODO: reduceRight
+    ///
+    /// Applies `f` to a start value `s` and all elements in `a` going from left to right.
+    ///
+    /// That is, the result is of the form: `f(...f(f(s, a[0]), a[1])..., xn)`.
+    ///
+    pub def foldLeft(f: (b, a) -> b & e, s: b, v: MutList[a]): b & Impure =
+        let len = length(v);
+        foldLeftHelper(f, s, v, len, 0)
 
-    // TODO: Additional operations from Array.
+    ///
+    /// Helper function for `foldLeft`.
+    ///
+    def foldLeftHelper(f: (b, a) -> b & e, s: b, v: MutList[a], len: Int32, i: Int32): b & Impure =
+        if (i < len)
+            let s1 = f(s, get(i, v)) as & Impure;
+            foldLeftHelper(f, s1, v, len, i + 1)
+        else
+            s
+
+    ///
+    /// Applies `f` to a start value `s` and all elements in `a` going from left to right.
+    ///
+    /// That is, the result is of the form: `f(...f(f(s, a[0]), a[1])..., xn)`.
+    ///
+    pub def foldRight(f: (a, b) -> b & e, s: b, v: MutList[a]): b & Impure =
+        let len = length(v);
+        foldRightHelper(f, s, v, len - 1)
+
+    ///
+    /// Helper function for `foldRight`.
+    ///
+    def foldRightHelper(f: (a, b) -> b & e, s: b, v: MutList[a], i: Int32): b & Impure =
+        if (i < 0)
+            s
+        else
+            let s1 = f(get(i, v), s) as & Impure;
+            foldRightHelper(f, s1, v, i - 1)
+
+    ///
+    /// Alias for `reduceLeft`.
+    ///
+    pub def reduce(f: (a, a) -> a & e, v: MutList[a]): Option[a] & Impure = reduceLeft(f, v)
+
+    ///
+    /// Applies `f` to all elements in `v` going from left to right until a single value `v` is obtained. Returns `Some(v)`.
+    ///
+    /// Returns `None` if `v` is empty.
+    ///
+    pub def reduceLeft(f: (a, a) -> a & e, v: MutList[a]): Option[a] & Impure = foldLeft((acc, x) -> match acc {
+        case Some(y) => Some(f(y, x))
+        case None => Some(x)
+    }, None, v)
 
-    /*
-     * Mutating operations.
-     */
+    ///
+    /// Applies `f` to all elements in `v` going from right to left until a single value `v` is obtained. Returns `Some(v)`.
+    ///
+    /// Returns `None` if `v` is empty.
+    ///
+    pub def reduceRight(f: (a, a) -> a & e, v: MutList[a]): Option[a] & Impure = foldRight((x, acc) -> match acc {
+        case Some(y) => Some(f(x, y))
+        case None => Some(x)
+    }, None, v)
 
     ///
     /// Removes all elements from the given mutable list `v`.
     ///
     /// Does not shrink its capacity.
     ///
-    pub def clear(v: mut MutList[a]): Unit = ???
+    pub def clear(v: mut MutList[a]): Unit & Impure =
+        let MutList(_, l) = v;
+        l := 0
```

Comment:
So, if we dont touch the array, we might hang on to values for a very long time.

I would also zero the array. So effectively this is almost like using `new()`. 

Later, once we have some kind of default value/null, we can probably improve this, so that we zero the array instead.